crates/astria-sequencer/src/accounts/ics20_transfer.rs

```diff
@@ -0,0 +1,349 @@
+use std::str::FromStr;
+
+use anyhow::{
+    Context as _,
+    Result,
+};
+use ibc_types::{
+    core::channel::{
+        channel,
+        msgs::{
+            MsgAcknowledgement,
+            MsgChannelCloseConfirm,
+            MsgChannelCloseInit,
+            MsgChannelOpenAck,
+            MsgChannelOpenConfirm,
+            MsgChannelOpenInit,
+            MsgChannelOpenTry,
+            MsgRecvPacket,
+            MsgTimeout,
+        },
+        ChannelId,
+        PortId,
+    },
+    transfer::acknowledgement::TokenTransferAcknowledgement,
+};
+use penumbra_ibc::component::app_handler::{
+    AppHandler,
+    AppHandlerCheck,
+    AppHandlerExecute,
+};
+use penumbra_proto::penumbra::core::component::ibc::v1alpha1::FungibleTokenPacketData;
+use penumbra_storage::{
+    StateRead,
+    StateWrite,
+};
+use proto::native::sequencer::v1alpha1::{
+    asset::{
+        IbcAsset,
+        Id,
+    },
+    Address,
+};
+
+use super::state_ext::{
+    StateReadExt as _,
+    StateWriteExt,
+};
+
+/// The ICS20 transfer handler.
+///
+/// See [here](https://github.com/cosmos/ibc/blob/main/spec/app/ics-020-fungible-token-transfer/README.md)
+/// for the specification which this is based on.
+#[derive(Clone)]
+pub(crate) struct Ics20Transfer;
+
+#[async_trait::async_trait]
+impl AppHandlerCheck for Ics20Transfer {
```

Comment:
So this is where the magic is: `AppHandlerCheck` must be impl'ed to make `with_handler` work.

Can we get that as a module doc comment?